lib/task.js

```diff
@@ -128,10 +128,46 @@ Task.prototype.chain = function _chain(f) {
  * @summary @Task[α, (β → γ)] => Task[α, β] → Task[α, γ]
  */
 
-Task.prototype.ap = function _ap(f2) {
-  return this.chain(function(f) {
-    return f2.map(f);
-  });
+Task.prototype.ap = function _ap(that) {
+  var forkThis = this.fork;
+  var forkThat = that.fork;
+  var cleanupThis = this.cleanup;
+  var cleanupThat = that.cleanup;
+
+  function cleanupBoth(state) {
+    cleanupThis(state[0]);
+    cleanupThat(state[1]);
+  }
+
+  return new Task(function(reject, resolve) {
+    var func, funcLoaded = false;
+    var val, valLoaded = false;
+    var allState;
+
+    var thisState = forkThis(reject, guard(function(x) {
```

Comment:
`reject` needs to check if the Task has already been rejected here (and in `forkThat`) or it could get called twice, in case both this and the applicative argument fails.